### PR TITLE
Hws patches

### PR DESCRIPTION
**Pull request by hamdanigbaria**

### CODE REVIEW

File: hws/src/dr_common.py [hunk 1]
```diff
@@ -381,7 +381,11 @@ def call_resource_dump(dev, dev_name, segment, index1, num_of_obj1, num_of_obj2,
 }
 
 
+FW_VERSION_MAJOR_CX6 = 22
 FW_VERSION_MAJOR_CX7 = 28
+
+EXTRA_HW_RES_FW_VERSION_MINOR = "39.0230"
+
```
Comment (kliteyn) [inline hws/src/dr_common.py:388]: It's better to have related names with the same prefix.
At some point you will have feature X starting with some other version.
Change it to FW_VERSION_MINOR_EXTRA_...
Then you have future-proofing for naming:
 - FW_VERSION_MINOR_X_...
 - FW_VERSION_MINOR_Y_...

File: hws/src/dr_hl.py [hunk 2]
```diff
@@ -455,18 +455,18 @@ def dr_hl_flex_parser_parser(hl_index, mask):
...
+                 [('regc_8', 32)],
+                 [('regc_9', 32)],
+                 [('regc_6', 32)],
+                 [('regc_7', 32)],
+                 [('regc_4', 32)],
+                 [('regc_5', 32)],
+                 [('regc_2', 32)],
+                 [('regc_3', 32)],
+                 [('regc_0', 32)],
+                 [('regc_1', 32)]
```
Comment (kliteyn) [inline hws/src/dr_hl.py:469]: Don't see the point for this...
But I'm not against it.

File: hws/src/dr_common.py (hunk 1 above)
Comment (hamdanigbaria) [reply hws/src/dr_common.py:388]: Done

File: hws/src/dr_hl.py (hunk 2 above)
Comment (hamdanigbaria) [reply hws/src/dr_hl.py:469]: Alex asked for this change, since users are familiar with this naming "regc_x" more

File: hws/src/dr_action.py [hunk 3]
```diff
@@ -3,170 +3,64 @@
...
-
-    if res != None and len(res) > 0:
-        action["field"] = res[0][0]
-    else:
-        action["dw_offset"] = dw_offset
-
-    action["length"] = length
-    action["left_shifter"] = left_shifter
-    action["inline_data"] = int(action_dw_1, 2)
+    action = dr_parse_set_action(action_dw_0, action_dw_1)
```
Comment (kliteyn) [inline hws/src/dr_action.py:23]: where is this function?

File: hws/src/dr_action.py [hunk 4]
```diff
@@ -3,170 +3,64 @@
...
-    mask = hex_to_bin_str(hex(int(length * "1", 2) << (32 - src_right_shifter)), 32)
-    res = get_field(src_dw_offset, mask)
-
-    if res != None and len(res) > 0:
-        action["src_field"] = res[0][0]
-    else:
-        action["src_dw_offset"] = src_dw_offset
-
-    action["src_right_shifter"] = src_right_shifter
+    action = dr_parse_copy_action(action_dw_0, action_dw_1)
```
Comment (kliteyn) [inline hws/src/dr_action.py:16]: where is this function?

File: hws/src/dr_action.py [hunk 5]
```diff
@@ -3,170 +3,64 @@
...
-
-    if res != None and len(res) > 0:
-        action["field"] = res[0][0]
-    else:
-        action["dw_offset"] = dw_offset
-
-    action["length"] = length
-    action["left_shifter"] = left_shifter
-    action["value"] = int(action_dw_1, 2)
+    action = dr_parse_add_action(action_dw_0, action_dw_1)
```
Comment (kliteyn) [inline hws/src/dr_action.py:30]: where is this function?

File: hws/src/dr_action.py [hunk 6]
```diff
@@ -3,170 +3,64 @@
...
     action_dw_0 = action_arr[index]
     action_dw_1 = action_arr[index + 1]
-    action = {"type" : "Remove by size"}
-    start_anchor = int(action_dw_0[10 : 16], 2)
-    field = modify_pattern_anchor_dic.get(start_anchor)
-    action["start_anchor"] = field if field != None else start_anchor
-    action["outer_l4_removed"] = int(action_dw_0[16 : 17], 2)
-    action["start_offset"] = int(action_dw_0[18 : 25], 2)
-    action["size"] = int(action_dw_0[26 : 32], 2)
+    action = dr_parse_remove_by_size_action(action_dw_0, action_dw_1)
```
Comment (kliteyn) [inline hws/src/dr_action.py:37]: and this

File: hws/src/dr_action.py [hunk 7]
```diff
@@ -3,170 +3,64 @@
...
-    start_anchor = int(action_dw_0[10 : 16], 2)
-    end_anchor = int(action_dw_0[18 : 24], 2)
-    field = modify_pattern_anchor_dic.get(start_anchor)
-    action["start_anchor"] = field if field != None else start_anchor
-    field = modify_pattern_anchor_dic.get(end_anchor)
-    action["end_anchor"] = field if field != None else end_anchor
-    action["decap"] = int(action_dw_0[28 : 29], 2)
-    action["vni_to_cqe"] = int(action_dw_0[29 : 30], 2)
-    action["qos_profile "] = int(action_dw_0[30 : 32], 2)
+    action = dr_parse_remove_header2header_action(action_dw_0, action_dw_1)
```
Comment (kliteyn) [inline hws/src/dr_action.py:44]: and this

File: hws/src/dr_action.py [hunk 8]
```diff
@@ -3,170 +3,64 @@
...
     action_dw_1 = action_arr[index + 1]
-    action = {"type" : "Insert inline"}
-    start_anchor = int(action_dw_0[10 : 16], 2)
-    field = modify_pattern_anchor_dic.get(start_anchor)
-    action["start_anchor"] = field if field != None else start_anchor
-    end_anchor = int(action_dw_0[18 : 24], 2)
-    field = modify_pattern_anchor_dic.get(end_anchor)
-    action["end_anchor"] = field if field != None else end_anchor
-    action["insert_data_inline"] = int(action_dw_1[0 : 32], 2)
+    action = dr_parse_insert_inline_action(action_dw_0, action_dw_1)
```
Comment (kliteyn) [inline hws/src/dr_action.py:51]: where is this function?

File: hws/src/dr_action.py [hunk 9]
```diff
@@ -3,170 +3,64 @@
...
-    start_anchor = int(action_dw_0[10 : 16], 2)
-    field = modify_pattern_anchor_dic.get(start_anchor)
-    action["start_anchor"] = field if field != None else start_anchor
-    end_anchor = int(action_dw_0[18 : 24], 2)
-    field = modify_pattern_anchor_dic.get(end_anchor)
-    action["end_anchor"] = field if field != None else end_anchor
-    action["size"] = int(action_dw_0[24 : 29], 2)
-    action["attributes"] = int(action_dw_0[29 : 32], 2)
-    action["pointer"] = int(action_dw_1[0 : 32], 2)
+    action = dr_parse_insert_by_pointer_action(action_dw_0, action_dw_1)
```
Comment (kliteyn) [inline hws/src/dr_action.py:58]: where is this function?

Comment (kliteyn) [inline hws/src/dr_action.py:58]: All this refactoring is nice. It has nothing to do though with pattern/args parsing.
You missed adding all those functions - please add them and do all this in a separate patch.

File: hws/src/dr_hw_resources.py [hunk 10]
```diff
@@ -136,104 +138,95 @@ def load_to_db(self):
...
-    length = int(raw[6:8], 16) & 0x1f
-    length = "0x20" if length == 0 else hex(length)
-    text = "SET: field: %s offset: %s length: %s" %\
-            (field, offset, length)
+def dr_parse_modify_pattern_set(raw):
+    _raw = hex_to_bin_str(raw, 64)
+    action = dr_parse_set_action(_raw[0:32], _raw[32:64], False)
 
-    return {"type": 0x1, "text": text}
+    return {"type": 0x1, "text": action_pretiffy(action, False), "raw": raw[0:8]}
```
Comment (kliteyn) [inline hws/src/dr_hw_resources.py:161]: I don't understand... You removed action_pretiffy function from this file, do you import it somewhere?

Comment (hamdanigbaria) [reply hws/src/dr_hw_resources.py:161]: yes, i import at the begining of the file:
from src.dr_common_functions import *

File: hws/src/dr_action.py (hunk 3 above)
Comment (hamdanigbaria) [reply hws/src/dr_action.py:23]: hws/src/dr_common_functions.py

File: hws/src/dr_action.py (hunk 4 above)
Comment (hamdanigbaria) [reply hws/src/dr_action.py:16]: hws/src/dr_common_functions.py

File: hws/src/dr_action.py (hunk 5 above)
Comment (hamdanigbaria) [reply hws/src/dr_action.py:30]: hws/src/dr_common_functions.py

File: hws/src/dr_action.py (hunk 6 above)
Comment (hamdanigbaria) [reply hws/src/dr_action.py:37]: hws/src/dr_common_functions.py

File: hws/src/dr_action.py (hunk 7 above)
Comment (hamdanigbaria) [reply hws/src/dr_action.py:44]: hws/src/dr_common_functions.py

File: hws/src/dr_action.py (hunk 8 above)
Comment (hamdanigbaria) [reply hws/src/dr_action.py:51]: hws/src/dr_common_functions.py

File: hws/src/dr_action.py (hunk 9 above)
Comment (hamdanigbaria) [reply hws/src/dr_action.py:58]: hws/src/dr_common_functions.py

Comment (hamdanigbaria) [reply hws/src/dr_action.py:58]: i have changed the parsing function to be used in the regular copy/set/add/... actions when set in the STE directly, and to also be used when copy/set/add/... are part of pattern and argument.
i have but the functions in hws/src/dr_common_functions.py

File: hws/src/dr_action.py (hunk 3 above)
Comment (hamdanigbaria) [reply hws/src/dr_action.py:23]: Pushed

**Review by kliteyn — APPROVED** [review]
